Build failure due to incompatible-pointer-types: `const fftw_r2r_kind *` vs `int *` for `fftw_plan_guru_r2r`

Hi. I tried to build #370 by myself but failed on EndeavourOS, where

- GCC 14.1.1
- python 3.12.3

The messages are as follows:

```ShellSession
$ pip install .
…
      Compiling pyfftw/pyfftw.pyx because it depends on /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/Cython/Includes/libc/string.pxd.
      [1/1] Cythonizing pyfftw/pyfftw.pyx
      INFO:root:building 'pyfftw.pyfftw' extension
      INFO:root:gcc -fno-strict-overflow -DNDEBUG -g -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -ffat-lto-objects -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -fPIC -DPYFFTW_HAVE_DOUBLE=1 -DPYFFTW_HAVE_DOUBLE_OMP=1 -DPYFFTW_HAVE_DOUBLE_THREADS=0 -DPYFFTW_HAVE_DOUBLE_MULTITHREADING=1 -DPYFFTW_HAVE_DOUBLE_MPI=0 -DPYFFTW_HAVE_SINGLE=1 -DPYFFTW_HAVE_SINGLE_OMP=1 -DPYFFTW_HAVE_SINGLE_THREADS=0 -DPYFFTW_HAVE_SINGLE_MULTITHREADING=1 -DPYFFTW_HAVE_SINGLE_MPI=0 -DPYFFTW_HAVE_LONG=1 -DPYFFTW_HAVE_LONG_OMP=1 -DPYFFTW_HAVE_LONG_THREADS=0 -DPYFFTW_HAVE_LONG_MULTITHREADING=1 -DPYFFTW_HAVE_LONG_MPI=0 -DPYFFTW_HAVE_MPI=0 -I./venv/include -I/usr/include/python3.12 -I./include -I./pyfftw -I/tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include -I./venv/include -I/usr/include -I./venv/include -I/usr/include/python3.12 -c pyfftw/pyfftw.c -o build/temp.linux-x86_64-cpython-312/pyfftw/pyfftw.o
      In file included from /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/ndarraytypes.h:1929,
                       from /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                       from /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/arrayobject.h:5,
                       from pyfftw/pyfftw.c:1306:
      /tmp/pip-build-env-kmgmq1nv/overlay/lib/python3.12/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
         17 | #warning "Using deprecated NumPy API, disable it with " \
            |  ^~~~~~~
      pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftw_plan_guru_r2r’:
      pyfftw/pyfftw.c:13236:164: error: passing argument 7 of ‘fftw_plan_guru_r2r’ from incompatible pointer type [-Wincompatible-pointer-types]
      13236 |   __pyx_r = ((void *)fftw_plan_guru_r2r(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((double *)__pyx_v__in), ((double *)__pyx_v__out), __pyx_v_direction, __pyx_v_flags));
            |                                                                                                                                                                    ^~~~~~~~~~~~~~~~~
            |                                                                                                                                                                    |
            |                                                                                                                                                                    int *
      In file included from pyfftw/pyfftw.c:1313:
      /usr/include/fftw3.h:465:1: note: expected ‘const fftw_r2r_kind *’ {aka ‘const enum fftw_r2r_kind_do_not_use_me *’} but argument is of type ‘int *’
        465 | FFTW_DEFINE_API(FFTW_MANGLE_DOUBLE, double, fftw_complex)
            | ^~~~~~~~~~~~~~~
      pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftwf_plan_guru_r2r’:
      pyfftw/pyfftw.c:13372:163: error: passing argument 7 of ‘fftwf_plan_guru_r2r’ from incompatible pointer type [-Wincompatible-pointer-types]
      13372 |   __pyx_r = ((void *)fftwf_plan_guru_r2r(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((float *)__pyx_v__in), ((float *)__pyx_v__out), __pyx_v_direction, __pyx_v_flags));
            |                                                                                                                                                                   ^~~~~~~~~~~~~~~~~
            |                                                                                                                                                                   |
            |                                                                                                                                                                   int *
      /usr/include/fftw3.h:466:1: note: expected ‘const fftwf_r2r_kind *’ {aka ‘const enum fftw_r2r_kind_do_not_use_me *’} but argument is of type ‘int *’
        466 | FFTW_DEFINE_API(FFTW_MANGLE_FLOAT, float, fftwf_complex)
            | ^~~~~~~~~~~~~~~
      pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftwl_plan_guru_r2r’:
      pyfftw/pyfftw.c:13508:175: error: passing argument 7 of ‘fftwl_plan_guru_r2r’ from incompatible pointer type [-Wincompatible-pointer-types]
      13508 |   __pyx_r = ((void *)fftwl_plan_guru_r2r(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((long double *)__pyx_v__in), ((long double *)__pyx_v__out), __pyx_v_direction, __pyx_v_flags));
            |                                                                                                                                                                               ^~~~~~~~~~~~~~~~~
            |                                                                                                                                                                               |
            |                                                                                                                                                                               int *
      /usr/include/fftw3.h:467:1: note: expected ‘const fftwl_r2r_kind *’ {aka ‘const enum fftw_r2r_kind_do_not_use_me *’} but argument is of type ‘int *’
        467 | FFTW_DEFINE_API(FFTW_MANGLE_LONG_DOUBLE, long double, fftwl_complex)
            | ^~~~~~~~~~~~~~~
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyFFTW
Failed to build pyFFTW
ERROR: Could not build wheels for pyFFTW, which is required to install pyproject.toml-based projects
```

GCC 8.5.0 (Rocky 8.9) actually issues the same warning—`int*` and `fftw_r2r_kind*` etc are incompatible—but exits with zero.

Solution would be:

* pass `-Wno-incompatible-pointer-types`
* fix the signatures of `fftwl_plan_guru_r2r`, `fftw_plan_guru_r2r` and  `fftwl_plan_guru_r2r`

> * fix the signatures of `fftwl_plan_guru_r2r`, `fftw_plan_guru_r2r` and  `fftwl_plan_guru_r2r`

I tried 

```diff
diff -ur pyFFTW-0.14.0/pyfftw/pyfftw.pxd pyFFTW-0.14.0.patched/pyfftw/pyfftw.pxd
--- pyFFTW-0.14.0/pyfftw/pyfftw.pxd	2024-07-10 23:02:53.000000000 +0200
+++ pyFFTW-0.14.0.patched/pyfftw/pyfftw.pxd	2024-09-05 13:08:02.759273313 +0200
@@ -179,6 +179,9 @@
     ctypedef struct fftw_iodim:
         pass
 
+    ctypedef enum fftw_r2r_kind:
+        pass
+
     # Double precision complex planner
     fftw_plan fftw_plan_guru_dft(
             int rank, fftw_iodim *dims,
@@ -247,21 +250,21 @@
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             double *_in, double *_out,
-            int *kind, unsigned flags) nogil
+            fftw_r2r_kind *kind, unsigned flags) nogil
 
     # Single precision real planner
     fftwf_plan fftwf_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             float *_in, float *_out,
-            int *kind, unsigned flags) nogil
+            fftw_r2r_kind *kind, unsigned flags) nogil
 
     # Long double precision real planner
     fftwl_plan fftwl_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             long double *_in, long double *_out,
-            int *kind, unsigned flags) nogil
+            fftw_r2r_kind *kind, unsigned flags) nogil
 
     # Double precision complex new array execute
     void fftw_execute_dft(fftw_plan,
diff -ur pyFFTW-0.14.0/pyfftw/pyfftw.pyx pyFFTW-0.14.0.patched/pyfftw/pyfftw.pyx
--- pyFFTW-0.14.0/pyfftw/pyfftw.pyx	2024-07-10 23:02:53.000000000 +0200
+++ pyFFTW-0.14.0.patched/pyfftw/pyfftw.pyx	2024-09-05 13:07:43.055321479 +0200
@@ -191,7 +191,7 @@
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             void *_in, void *_out,
-            int *direction, int flags) noexcept nogil:
+            fftw_r2r_kind *direction, int flags) noexcept nogil:
 
     return <void *>fftw_plan_guru_r2r(rank, dims,
             howmany_rank, howmany_dims,
@@ -239,7 +239,7 @@
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             void *_in, void *_out,
-            int *direction, int flags) noexcept nogil:
+            fftw_r2r_kind *direction, int flags) noexcept nogil:
 
     return <void *>fftwf_plan_guru_r2r(rank, dims,
             howmany_rank, howmany_dims,
@@ -287,7 +287,7 @@
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             void *_in, void *_out,
-            int *direction, int flags) noexcept nogil:
+            fftw_r2r_kind *direction, int flags) noexcept nogil:
 
     return <void *>fftwl_plan_guru_r2r(rank, dims,
             howmany_rank, howmany_dims,
```

But then there are more incompatible pointers pertaining complex double. I am not versed enough in C/Cython to fix this.

```C
[   10s]   INFO:root:gcc -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -g -DOPENSSL_LOAD_CONF -fwrapv -fno-semantic-interposition -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -g -IVendor/ -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -g -IVendor/ -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fPIC -DPYFFTW_HAVE_DOUBLE=1 -DPYFFTW_HAVE_DOUBLE_OMP=0 -DPYFFTW_HAVE_DOUBLE_THREADS=1 -DPYFFTW_HAVE_DOUBLE_MULTITHREADING=1 -DPYFFTW_HAVE_DOUBLE_MPI=0 -DPYFFTW_HAVE_SINGLE=1 -DPYFFTW_HAVE_SINGLE_OMP=0 -DPYFFTW_HAVE_SINGLE_THREADS=1 -DPYFFTW_HAVE_SINGLE_MULTITHREADING=1 -DPYFFTW_HAVE_SINGLE_MPI=0 -DPYFFTW_HAVE_LONG=1 -DPYFFTW_HAVE_LONG_OMP=0 -DPYFFTW_HAVE_LONG_THREADS=1 -DPYFFTW_HAVE_LONG_MULTITHREADING=1 -DPYFFTW_HAVE_LONG_MPI=0 -DPYFFTW_HAVE_MPI=0 -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -I/home/abuild/rpmbuild/BUILD/pyFFTW-0.14.0/include -I/home/abuild/rpmbuild/BUILD/pyFFTW-0.14.0/pyfftw -I/usr/lib64/python3.10/site-packages/numpy/_core/include -I/usr/include -I/usr/include -I/usr/include/python3.10 -I/usr/include/python3.10 -c pyfftw/pyfftw.c -o build/temp.linux-x86_64-cpython-310/pyfftw/pyfftw.o
[   10s]   pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw__fftw_plan_guru_dft’:
[   10s]   pyfftw/pyfftw.c:13652:114: error: passing argument 5 of ‘fftw_plan_guru_dft’ from incompatible pointer type [-Wincompatible-pointer-types]
[   10s]   13652 |   __pyx_r = ((void *)fftw_plan_guru_dft(__pyx_v_rank, __pyx_v_dims, __pyx_v_howmany_rank, __pyx_v_howmany_dims, ((cdouble *)__pyx_v__in), ((cdouble *)__pyx_v__out), (__pyx_v_direction[0]), __pyx_v_flags));
[   10s]         |                                                                                                                 ~^~~~~~~~~~~~~~~~~~~~~~~
[   10s]         |                                                                                                                  |
[   10s]         |                                                                                                                  double (*)[2]
[   10s]   In file included from pyfftw/pyfftw.c:1335:
[   10s]   /usr/include/fftw3.h:465:1: note: expected ‘_Complex double *’ but argument is of type ‘double (*)[2]’
[   10s]     465 | FFTW_DEFINE_API(FFTW_MANGLE_DOUBLE, double, fftw_complex)
[   10s]         | ^
...
```

--- a/pyfftw/pyfftw.pxd
+++ b/pyfftw/pyfftw.pxd
@@ -247,21 +247,21 @@ cdef extern from 'fftw3.h':
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             double *_in, double *_out,
-            int *kind, unsigned flags) nogil
+            fftw_r2r_kind *kind, unsigned flags) nogil
 
     # Single precision real planner
     fftwf_plan fftwf_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             float *_in, float *_out,
-            int *kind, unsigned flags) nogil
+            fftwf_r2r_kind *kind, unsigned flags) nogil
 
     # Long double precision real planner
     fftwl_plan fftwl_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             long double *_in, long double *_out,
-            int *kind, unsigned flags) nogil
+            fftwl_r2r_kind *kind, unsigned flags) nogil
 
     # Double precision complex new array execute
     void fftw_execute_dft(fftw_plan,
@@ -373,6 +373,48 @@ cdef extern from 'fftw3.h':
 
     double FFTW_NO_TIMELIMIT
 
+    ctypedef enum fftw_r2r_kind:
+        FFTW_R2HC = 0
+        FFTW_HC2R = 1
+        FFTW_DHT = 2
+        # from fftw3.f 3.3.3; may not be valid for different versions of FFTW.
+        FFTW_REDFT00  = 3
+        FFTW_REDFT01  = 4
+        FFTW_REDFT10  = 5
+        FFTW_REDFT11  = 6
+        FFTW_RODFT00  = 7
+        FFTW_RODFT01  = 8
+        FFTW_RODFT10  = 9
+        FFTW_RODFT11  = 10
+
+    ctypedef enum fftwf_r2r_kind:
+        FFTW_R2HC = 0
+        FFTW_HC2R = 1
+        FFTW_DHT = 2
+        # from fftw3.f 3.3.3; may not be valid for different versions of FFTW.
+        FFTW_REDFT00  = 3
+        FFTW_REDFT01  = 4
+        FFTW_REDFT10  = 5
+        FFTW_REDFT11  = 6
+        FFTW_RODFT00  = 7
+        FFTW_RODFT01  = 8
+        FFTW_RODFT10  = 9
+        FFTW_RODFT11  = 10
+
+    ctypedef enum fftwl_r2r_kind:
+        FFTW_R2HC = 0
+        FFTW_HC2R = 1
+        FFTW_DHT = 2
+        # from fftw3.f 3.3.3; may not be valid for different versions of FFTW.
+        FFTW_REDFT00  = 3
+        FFTW_REDFT01  = 4
+        FFTW_REDFT10  = 5
+        FFTW_REDFT11  = 6
+        FFTW_RODFT00  = 7
+        FFTW_RODFT01  = 8
+        FFTW_RODFT10  = 9
+        FFTW_RODFT11  = 10
+
 # Define function pointers that can act as a placeholder
 # for whichever dtype is used (the problem being that fftw
 # has different function names and signatures for all the
@@ -408,15 +450,6 @@ ctypedef bint (*validator)(np.ndarray input_array,
 cdef enum:
     FFTW_FORWARD = -1
     FFTW_BACKWARD = 1
-    # from fftw3.f 3.3.3; may not be valid for different versions of FFTW.
-    FFTW_REDFT00  = 3
-    FFTW_REDFT01  = 4
-    FFTW_REDFT10  = 5
-    FFTW_REDFT11  = 6
-    FFTW_RODFT00  = 7
-    FFTW_RODFT01  = 8
-    FFTW_RODFT10  = 9
-    FFTW_RODFT11  = 10
 
 # Documented flags
 cdef enum:

--- a/pyfftw/pyfftw.pyx
+++ b/pyfftw/pyfftw.pyx
@@ -191,12 +191,12 @@ cdef void* _fftw_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             void *_in, void *_out,
-            int *direction, int flags) noexcept nogil:
+            fftw_r2r_kind *direction, int flags) noexcept nogil:
 
     return <void *>fftw_plan_guru_r2r(rank, dims,
             howmany_rank, howmany_dims,
             <double *>_in, <double *>_out,
-            direction, flags)
+            <fftw_r2r_kind *>direction, flags)
 
 # Complex single precision
 cdef void* _fftwf_plan_guru_dft(
@@ -239,12 +239,12 @@ cdef void* _fftwf_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             void *_in, void *_out,
-            int *direction, int flags) noexcept nogil:
+            fftwf_r2r_kind *direction, int flags) noexcept nogil:
 
     return <void *>fftwf_plan_guru_r2r(rank, dims,
             howmany_rank, howmany_dims,
             <float *>_in, <float *>_out,
-            direction, flags)
+            <fftwf_r2r_kind *>direction, flags)
 
 # Complex long double precision
 cdef void* _fftwl_plan_guru_dft(
@@ -287,12 +287,12 @@ cdef void* _fftwl_plan_guru_r2r(
             int rank, fftw_iodim *dims,
             int howmany_rank, fftw_iodim *howmany_dims,
             void *_in, void *_out,
-            int *direction, int flags) noexcept nogil:
+            fftwl_r2r_kind *direction, int flags) noexcept nogil:
 
     return <void *>fftwl_plan_guru_r2r(rank, dims,
             howmany_rank, howmany_dims,
             <long double *>_in, <long double *>_out,
-            direction, flags)
+            <fftwl_r2r_kind *>direction, flags)
 
 #    Executors
 #    =========

